### ISSUE TITLE
[BUG] Auxiliary wire handling in `metric_tensor` and `hadamard_grad`

### ISSUE THREAD

**Issue by dwierichs** [PennyLaneAI/pennylane#4325]
### Expected behavior

The handling of the `aux_wire` and `device_wires` kwargs is the same for `transforms.metric_tensor` and `gradients.hadamard_grad`, and as many use cases as possible are supported.

### Actual behavior

Differences between the two functions:
1. `metric_tensor` can handle an `aux_wire` input without a `device_wires` input, `hadamard_grad` raises an error (which is not necessary imo).
2. `hadamard_grad` can handle `aux_wire` to be a `Wires` object, `metric_tensor` can not.
3. `metric_tensor` can handle `aux_wire` to be a multi-character string or a simple integer, `hadamard_grad` can not.
4. `hadamard_grad` correctly detects reusage of a wire from the tape, `metric_tensor` does not (and produces flawed circuits)

### Additional information

This can easily be fixed by unifying the handling of `aux_wire` and `device_wires` from the two functions and cherrypicking the best of both worlds.

### Source code

Take the circuit
```python
dev = qml.device("default.qubit", wires=[0, 1, 2, "aux"])

@qml.qnode(dev, interface="jax")
def circuit(x, argnums=None):
    qml.RX(x[0], 0)
    qml.CNOT([0, 1])
    qml.RX(x[1], 0)
    return qml.expval(Z(0))

x = jnp.array([1.3, 0.2])
```
Then
```python
# 1
qml.gradients.hadamard_grad(circuit, aux_wire=qml.wires.Wires(2))(x) # does not work, None not subscriptable
# 2
qml.transforms.metric_tensor(circuit, aux_wire=qml.wires.Wires(2))(x) # does not work, wire not recognized
# 3
qml.gradients.hadamard_grad(circuit, aux_wire=2)(x) # does not work, int not subscriptable
qml.gradients.hadamard_grad(circuit, aux_wire="aux")(x) # does not work, wire not found
# 4
qml.transforms.metric_tensor(circuit, aux_wire=1)(x) # runs through but creates invalid tapes, reusing tape wire.
qml.transforms.metric_tensor(circuit, aux_wire=0)(x) #raises not-so-useful error
```

### System information

```shell
pl dev
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.
